numpy/random/_generator.pyi

```diff
@@ -0,0 +1,394 @@
+    @overload
+    def permutation(self, x: ArrayLike, axis: int = ...) -> ndarray[Any, Any]: ...
+    @overload
+    def standard_cauchy(self, size: None = ...) -> float: ...  # type: ignore[misc]
+    @overload
+    def standard_cauchy(self, size: Optional[_ShapeLike] = ...) -> ndarray[Any, dtype[float64]]: ...
+    @overload
+    def standard_exponential(  # type: ignore[misc]
+        self,
+        size: None = ...,
+        dtype: Union[_DTypeLikeFloat32, _DTypeLikeFloat64] = ...,
+        method: Literal["zig", "inv"] = ...,
+        out: None = ...,
+    ) -> float: ...
+    @overload
+    def standard_exponential(
+        self,
+        size: _ShapeLike = ...,
+        dtype: Union[_DTypeLikeFloat32, _DTypeLikeFloat64] = ...,
+        method: Literal["zig", "inv"] = ...,
+        out: Optional[ndarray[Any, dtype[Union[float32, float64]]]] = ...,
+    ) -> ndarray[Any, dtype[Union[float32, float64]]]: ...
+    @overload
+    def standard_exponential(
+        self,
+        size: Optional[_ShapeLike] = ...,
+        dtype: Union[_DTypeLikeFloat32, _DTypeLikeFloat64] = ...,
+        method: Literal["zig", "inv"] = ...,
+        out: ndarray[Any, dtype[Union[float32, float64]]] = ...,
+    ) -> ndarray[Any, dtype[Union[float32, float64]]]: ...
+    @overload
+    def random(  # type: ignore[misc]
+        self,
+        size: None = ...,
+        dtype: Union[_DTypeLikeFloat32, _DTypeLikeFloat64] = ...,
+        out: None = ...,
+    ) -> float: ...
+    @overload
+    def random(
+        self,
+        size: _ShapeLike = ...,
+        dtype: Union[_DTypeLikeFloat32, _DTypeLikeFloat64] = ...,
+        out: Optional[ndarray[Any, dtype[Union[float32, float64]]]] = ...,
+    ) -> ndarray[Any, dtype[Union[float32, float64]]]: ...
+    @overload
+    def beta(self, a: float, b: float, size: None = ...) -> float: ...  # type: ignore[misc]
+    @overload
+    def beta(
+        self, a: _ArrayLikeFloat_co, b: _ArrayLikeFloat_co, size: Optional[_ShapeLike] = ...
+    ) -> ndarray[Any, dtype[float64]]: ...
+    @overload
+    def exponential(self, scale: float = ..., size: None = ...) -> float: ...  # type: ignore[misc]
+    @overload
+    def exponential(
+        self, scale: _ArrayLikeFloat_co = ..., size: Optional[_ShapeLike] = ...
+    ) -> ndarray[Any, dtype[float64]]: ...
+    @overload
+    def integers(  # type: ignore[misc]
+        self,
+        low: int,
+        high: Optional[int] = ...,
+        size: None = ...,
+        dtype: Union[_DTypeLikeBool, _DTypeLikeInt, _DTypeLikeUInt] = ...,
+        endpoint: bool = ...,
+    ) -> int: ...
+    @overload
+    def integers(
+        self,
+        low: _ArrayLikeInt_co,
+        high: Optional[_ArrayLikeInt_co] = ...,
+        size: Optional[_ShapeLike] = ...,
+        dtype: Union[_DTypeLikeBool, _DTypeLikeInt, _DTypeLikeUInt] = ...,
+        endpoint: bool = ...,
+    ) -> ndarray[Any, dtype[integer]]: ...
```

Comment:
There is still an abstract dtype here that should be narrowed down (or set to `Any`).

Reply:
I made it specific (although it is literally any integer type bool & (u)int(8/16/32/64)).

Reply:
This still doesn't really resolve the issue, as now you're stuck with an unsafe union (one that can be avoided).

One thing to be mindful of is that `Union` represents a union and not an intersection,
_i.e._ any operation performed on the to-be returned array will now have to be compatible with both the
`np.bool_`-, `np.unsignedinteger`- and `np.signedinteger`-based dtypes.

If you don't want to add the additional overloads for narrowing down the return types, 
then the dtype should be set to `Any` here.

Reply:
I see.  I've added 11 overloads( (u)int8/16/32/16+uint+int_+bool).  Probably need a look over.

For them I used the recipe
```
Union[
dtype[_T],
Type[_T],
_TnameCodes,
_SupportsDType[dtype[_T]]
]
```

where I am loosely using `_T` as a type, e.g., `uint16` and `_Tname` is `_Uint16`.

I think I got the two "specials" correct, `Type[bool]` -> `ndarray[Any, bool_]` and `Type[int]` ->`ndarray[Any,int_]`

Reply:
Looks good overall, though it's important that `bool` comes _before_ the `int` overload (this will have to be changed).

`bool` is a sub-type of `int` (a subclass even), so with the current overload order
the likes of `integers(..., dtype=bool)` will be parsed by the `int`-based overload, 
as mypy uses a first-come-first-served principle for matching overloads.

Reply:
I've moved bool up and added yet another overload (that is more specific but not absolutely necessary since the output of the new is bool while the previous said it was an int).